plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/printing/JKCodeBuilder.kt

```diff
@@ -777,9 +790,13 @@ class JKCodeBuilder(context: NewJ2kConverterContext) {
 
         override fun visitKtPrimaryConstructorRaw(ktPrimaryConstructor: JKKtPrimaryConstructor) {
             ktPrimaryConstructor.annotationList.accept(this)
-            renderModifiersList(ktPrimaryConstructor)
+            val redundantConstructor = redundantModifiersForConstructor(ktPrimaryConstructor)
+            if (!redundantConstructor) {
+                renderModifiersList(ktPrimaryConstructor)
+            }
 
-            val needConstructorKeyword = ktPrimaryConstructor.hasAnnotations || ktPrimaryConstructor.visibility != PUBLIC
+            val needConstructorKeyword =
+                (ktPrimaryConstructor.hasAnnotations || ktPrimaryConstructor.visibility != PUBLIC) && !redundantConstructor
```

Comment:
`ktPrimaryConstructor.visibility != PUBLIC` should probably be moved out of this place now